### ISSUE TITLE
kd loss under tensor parallelism trains with gradients scaled by tp_size

### ISSUE THREAD

**Issue by stanley1208** [NVIDIA-NeMo/Automodel#3776]
training with knowledge distillation and a tensor parallel student gives gradients that are exactly tp_size times too large on the kd side. the loss you see in the logs is perfectly correct, so nothing ever looks wrong. but underneath, at tp=8 with kd_ratio=0.5, the model effectively trains with an 8:1 kd to ce mix instead of 1:1. your kd_ratio setting is silently overridden.

the cause is one line in `components/loss/kd_loss.py` (line 107). the kl sum across tp ranks uses the differentiable `torch.distributed.nn.functional.all_reduce`, whose backward all_reduces the gradients again. since every tp rank computes the same loss and every rank calls backward, that second all_reduce adds up tp identical gradients. each rank's gradient should be counted once, it gets counted tp times.

easy to see on cpu with 2 gloo ranks acting as tp=2, real KDLoss, compared against the plain single process KDLoss on the same inputs:

```
reference (non-TP) loss: 0.368836
rank 0: TP loss=0.368836 | grad/ref ratio = 2.0000
rank 1: TP loss=0.368836 | grad/ref ratio = 2.0000
```

same loss, every gradient exactly doubled. the same thing reproduces through real DTensor plumbing with the tp group auto detected: the ce gradients come out right (1x), the kd gradients come out 2x.

the fix is one line plus a tiny helper: that kl all_reduce should pass gradients through unchanged in backward, since each rank already holds the full gradient of the shared loss. important detail: the softmax denominator all_reduce two lines above must stay as is, that one really needs the gradients summed. verified with the same setup:

```
rank 0 fixed: ratio = 1.0000, max diff from reference = 1.1e-08
rank 1 fixed: ratio = 1.0000, max diff from reference = 7.5e-09
```

why the existing test didn't catch it: the kd tp correctness test checks that gradients exist, not what their values are.

happy to send a pr with the fix and a gradient parity test.

**Comment by mikemikimike** [comment]
Implemented the tensor-parallel KD gradient fix in #3777. The final KL aggregation now uses a non-autograd-aware reduction so each rank retains its local vocabulary gradient, with a two-rank regression test against the full-vocabulary reference.

**Comment by stanley1208** [comment]
thanks for the pr! 
I ran my gradient check from #3776 against your branch and it passes: the kd gradients now match the full vocab reference exactly (ratio 1.0, max diff 1e-08), on both tp paths. loss values unchanged. the fix works.

got one small thought here: the in-place all_reduce works today, but if the line producing kl_local ever changes to an op that saves its output (like exp), autograd will throw an error here. a tiny custom autograd function (forward: all_reduce, backward: pass through) would avoid that.

**Comment by mikemikimike** [comment]
Implemented the follow-up in #3777 (commit 12d4970). The final KL aggregation now uses a private autograd function that all-reduces a clone in forward and passes each rank's local gradient through in backward, so future producers that save their output for autograd cannot be invalidated by an in-place reduction. Added a regression test for that saved-tensor case.
